### PR TITLE
 #9 Externalize the json format

### PR DESCRIPTION
**Pull request by pepeiborra**
This pull request:
- adds a new datatype `DebugTrace` to represent the JSON schema, 
- adds a new function `getDebugTrace` to snapshot the trace, 
- adds `xxxTrace` variants to all the existing functions that consume the debug trace, and 
- fixes a minor bug with the rendering of the arguments of a function call in `debugPrint`.

### CODE REVIEW

File: debug.cabal [hunk 1]
```diff
@@ -59,5 +64,6 @@ test-suite debug-test
 
     build-depends:
         base == 4.*,
+        bytestring,
```
Comment (ndmitchell) [inline debug.cabal]: Is this required? Weeder thinks not.

File: debug.cabal [hunk 2]
```diff
@@ -33,15 +33,20 @@ library
...
+        aeson,
         ghc-prim,
         extra,
         containers,
         directory,
         template-haskell,
         open-browser,
+        text,
         uniplate,
+        unordered-containers,
```
Comment (ndmitchell) [inline debug.cabal]: Given the addition of unordered containers it probably makes sense to stop using containers

Comment (pepeiborra) [reply debug.cabal]: Agreed

File: debug.cabal (hunk 1 above)
Comment (pepeiborra) [reply debug.cabal]: It turns out it is required, to unpack the json bytestring produced by `encode` in the `debugJSON` body.

**Comment by pepeiborra** [comment]
Good to go !

**Comment by ndmitchell** [comment]
Awesome. Will merge when I next get to a computer.

**Pull request merged**